core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -135,19 +123,18 @@ public TableScan appendsAfter(long fromSnapshotId) {
 
   @Override
   public TableScan useSnapshot(long scanSnapshotId) {
-    Preconditions.checkArgument(this.snapshotId == null,
-        "Cannot override snapshot, already set to id=%s", snapshotId);
+    Preconditions.checkState(context.snapshotId() == null,
```

Comment:
I think this should continue to be an `IllegalArgumentException` because it is validating the argument passed into `useSnapshot`. The problem is not with the context, it is with how the user is trying to modify that context.